src/sqlfluff/rules/L009.py

```diff
@@ -1,13 +1,35 @@
 """Implementation of Rule L009."""
-from typing import Optional
-
-from sqlfluff.core.parser import NewlineSegment
+from typing import List, Optional, Tuple
 
+from sqlfluff.core.parser import BaseSegment, NewlineSegment
 from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
 from sqlfluff.core.rules.doc_decorators import document_fix_compatible
 from sqlfluff.core.rules.functional import Segments, sp, tsp
 
 
+def get_trailing_newlines(segment: BaseSegment) -> List[BaseSegment]:
+    """Returns list of trailing newlines in the tree."""
+    result = []
+    for seg in segment.recursive_crawl_all(reverse=True):
+        if seg.is_type("newline"):
+            result.append(seg)
+        if not seg.is_whitespace and not seg.is_type("dedent"):
+            break
+    return result
+
+
+def get_last_segment(segment: Segments) -> Tuple[List[BaseSegment], Segments]:
+    """Returns rightmost & lowest descendant and its "parent stack"."""
+    parent_stack: List[BaseSegment] = []
+    while True:
+        children = segment.children()
+        if children:
+            parent_stack.append(segment[0])
+            segment = children.last()
+        else:
+            return parent_stack, segment
```

Comment:
Why doesn't it use the new `final_segment` function?

Reply:
Because it also needs the parent stack, which is not provided by `RuleContext.final_segment`. I could've had `RuleContext` return both, but that'd be harder to document clearly, and AFAIK, this is the only rule that needs it.